### PR TITLE
Leak rate plotter

### PR DESCRIPTION
**Pull request by radicalvictorian**
This consists of both front and back end commits that allow finalqc leakdata to be viewed. I had issues getting the plotleakrate script to make the graph appear in the layout, so I implemented a button that can be pushed to make the graph pop up. I can attempt to make it appear in the window if that's what you desire. But that may be somewhat more time intensive than is desirable.

### CODE REVIEW

**Comment by radicalvictorian** [comment]
I put the leakdata acquisition function in panel_leak_utilities. I also eliminated function duplication and added the title. I also added the dropdown menu to select the tag. Additionally, I added back the previously removed finalqc windows and backend code.

**Comment by bamesserly** [comment]
[Add leak rate plots to the DBV](https://trello.com/c/pZiz9XL6/325-add-leak-rate-plots-to-the-dbv)

File: guis/panel/leak/panel_leak_utilities.py [hunk 1]
```diff
@@ -92,3 +93,94 @@ def get_time(time_str):
...
+        leak_df.rename(columns={"heater_pct":"Heater%"}, inplace=True)
+    
+        # drop other columns
+        leak_df.drop(columns=['id', 'trial'], axis=1, inplace=True)
+        
+        con.close()
+        
+        return leak_df
+    
+    except:
```
Comment (bamesserly) [inline guis/panel/leak/panel_leak_utilities.py:145]: Two complaints here:
1. not a fan of putting so much code underneath a `try:` statement. It's like you're not sure where you expect a problem to occur. Maybe there's a specific reason for putting so much in it, but I'm not seeing it. Another sneaker reason not to put too much code underneath a try: different languages might still try to process the lines after the exception and before the except (I don't think python does, so this is for other languages) which could produce bad behavior and at least it will be slow.
2. you should catch specific exceptions. looks like we want a sqlite exception of some kind. It's obviously not critical, but it's just good practice. if you tell me that you googled for 1 minute can couldn't find the right syntax for the exception, that's probably a good enough reason for me. And if you really want to catch ANY error and do something about it, you can do `except Exception as e: print(e)` to learn about it.

In summary: just put the one or two lines where the exception can occur underneath the try: and then catch a specific exception if you know what the problem is going to be.

File: guis/panel/leak/panel_leak_utilities.py [hunk 2]
```diff
@@ -92,3 +93,94 @@ def get_time(time_str):
...
+################################################################################
+# returns a list of leak test tags for an inputted panel
+################################################################################
+
+def getLeakTags(panel):
+    tag_list = []
+    con = sqlite3.connect('data/database.db')
+    cursor = con.cursor()
+    
+    # acquire straw location
```
Comment (bamesserly) [inline guis/panel/leak/panel_leak_utilities.py:155]: There are several functions lying around the repository that get a procedure ID and/or straw location. Should use those.

Comment (bamesserly) [reply guis/panel/leak/panel_leak_utilities.py:155]: I think you might have even worked with one or two of them here #147 or maybe #139?

If one doesn't exist in a common place, perhaps you could add it to `gui_utils.py`. I think that's a file.

File: guis/panel/leak/panel_leak_utilities.py [hunk 3]
```diff
@@ -92,3 +93,94 @@ def get_time(time_str):
...
         return df
+
+
+################################################################################
+# READ INPUT
+# Read raw data from the database into a dataframe, structurally equivalent to that of ReadLeakRateFile, but with timestamps retained
+# TIME(DAYS)    FILLPSIA    RefPSIA PRESSURE(PSI)   BOX TEMPERATURE(C)  ROOM TEMPERATURE(C) Heater%
+################################################################################
+
+def readLeakDb(panel, tag):
```
Comment (bamesserly) [inline guis/panel/leak/panel_leak_utilities.py:104]: I recognize that function naming case conventions are all over the place in this repo, but I have a personal and irrational distaste for `camelCase` used in any situation. For functions, it's either `snake_case` or `UpperCamelCase` for me.
Let's go with a slightly more descriptive name here: `ReadLeakDataFromDB`. And likewise, the other one should be changed to `ReadLeakDataFromFile`. I'd also accept `ReadRawLeakDataFromDB` or even `ReadRawPanelLeakDataFromDB`.

File: guis/panel/leak/panel_leak_utilities.py [hunk 4]
```diff
@@ -92,3 +93,94 @@ def get_time(time_str):
...
+    leak_df = pd.DataFrame()
+    
+    #ensure that panel is type str
+    panel = str(panel)
+    
+    
+    con = sqlite3.connect('data/database.db')
+    cursor = con.cursor()
+
+    # acquire straw location
```
Comment (bamesserly) [inline guis/panel/leak/panel_leak_utilities.py:114]: There are several functions lying around the repository that get a procedure ID and/or straw location. Should use those.

I think you might have even worked with one or two of them here #147 or maybe #139?

If one doesn't exist in a common place, perhaps you could add it to `gui_utils.py`. I think that's a file. At worst, you can add it here.

File: guis/panel/leak/panel_leak_utilities.py (hunk 2 above)
Comment (bamesserly) [inline guis/panel/leak/panel_leak_utilities.py:155]: Again, use a common function for this.

Additionally, as-is, this is unnecessary code duplication between this function and the above.

File: guis/panel/leak/PlotLeakRate.py [hunk 5]
```diff
@@ -231,9 +232,15 @@ def FitAndPlot_2(df, axDiffP):
 ################################################################################
 # main
 ################################################################################
-def main(options):
+def main(options, panel=None, tag=None):
     # read raw data files into a dataframe
-    df = ReadLeakRateFile(options.infile, options.is_new_format)
+    if isinstance(options, pd.DataFrame):
```
Comment (bamesserly) [inline guis/panel/leak/PlotLeakRate.py:237]: This overloading of `options` and comparison with `isinstance` is not good practice. It should be clear from a function's interface what types of objects you could pass. This would absolutely not fly in just about any other language.

See [here](https://stackoverflow.com/questions/3111611/is-this-use-of-isinstance-pythonic-good) for discussion.

It's coming from a good place, though. Indeed this function feels wrong, doesn't it? Originally, it was two functions, but that was bad because there was a ton of code duplication. Now that it's one function, we have too many if's and checks.

There's a way to solve this. Gotta think about it. It's in [here](https://en.wikipedia.org/wiki/Design_Patterns) somewhere...

Comment (bamesserly) [reply guis/panel/leak/PlotLeakRate.py:237]: Without getting fancy with design patterns, can we just shove everything into the options?

File: guis/panel/leak/PlotLeakRate.py [hunk 6]
```diff
@@ -257,12 +264,20 @@ def main(options):
 
     # get the start, end, and elapsed times over which to make the fit
     total_duration = df["TIME(DAYS)"].iat[-1]
-    fit_start_time = (
-        GetFitStartTime(total_duration)
-        if not options.fit_start_time
-        else options.fit_start_time
-    )
-    fit_end_time = total_duration if not options.fit_end_time else options.fit_end_time
+    if dbCall == False:
```
Comment (bamesserly) [inline guis/panel/leak/PlotLeakRate.py:267]: `if not dbCall`

File: guis/panel/leak/PlotLeakRate.py (hunk 5 above)
Comment (bamesserly) [reply guis/panel/leak/PlotLeakRate.py:237]: https://en.wikipedia.org/wiki/Strategy_pattern

**Comment by bamesserly** [comment]
I do believe Adam picked this up and finished it.
